### DISCUSSION TITLE
Creating Containers: Pip install conflicts

### DISCUSSION THREAD

**Discussion by stratus-ss** [erew123/alltalk_tts#79] [Dev Discussions]
I started to work on a Containerfile for the project as part of the backlog the project indicated, As part of this I noted that the packages that attempt to get installed as part of the nvidia package have conflicts

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pydantic-core 2.14.6 requires typing-extensions!=4.7.0,>=4.6.0, but you have typing-extensions 4.4.0 which is incompatible.
huggingface-hub 0.20.2 requires fsspec>=2023.5.0, but you have fsspec 2023.4.0 which is incompatible.
gruut 2.2.3 requires networkx<3.0.0,>=2.5.0, but you have networkx 3.0 which is incompatible.
tts 0.21.3 requires fsspec>=2023.6.0, but you have fsspec 2023.4.0 which is incompatible.
tts 0.21.3 requires numpy>=1.24.3; python_version > "3.10", but you have numpy 1.24.1 which is incompatible.
```

And so on. This causes issues because the container build process will die on any error. In my case, the container build process is dying on the TTS & numpy compatibility. So I guess my question is, how can I help resolve these dependecy errors? Does the project really need the exact versions specified (maybe a pip freeze that went sideways?)

If not, can the requirements be loosened?

**Comment by erew123** [comment]
Hiya

So this is the standalone version. First off, yes the requirements are a pain in the butt with this. Have a look at line 337 on here https://github.com/erew123/alltalk_tts/blob/main/atsetup.bat

Here is the difficulty I faced. If someone has a cached version of Pytorch and even then, if you pip clear the cache, just doing a `pip install torch==2.1.0 torchvision==0.16.0 torchaudio==2.1.0 --index-url https://download.pytorch.org/whl/cu121` this, doesn't install it with CUDA.

So you have to do `pip install torch==2.1.0 torchvision==0.16.0 torchaudio==2.1.0 --upgrade --force-reinstall --index-url https://download.pytorch.org/whl/cu121` which forces the cuda to be installed, but because you are forcing it, it downgrades/affects other bits. (Specifically `--upgrade --force-reinstall`)

Hence you will see in the script I force an upgrade to these parts. This is obviously separate from what's occurring when you manually run the `pip install -r xxxxxx_requirements.txt` files and for the most part, I've let that be led by the requirements for text-generation-webui. Whist slightly out of date, there is a breakdown here https://github.com/erew123/alltalk_tts/issues/23

Maybe the simplest approach would be to manually force your versions to upgrade with the following:

    pip install numpy==1.24.4
    pip install typing-extensions>=4.8.0
    pip install networkx<3.0.0,>=2.5.0
    pip install fsspec>=2023.6.0
    pip install soundfile==0.12.1
    pip install uvicorn==0.24.0.post1
    pip install TTS==0.21.3
    pip install fastapi==0.104.1
    pip install Jinja2==3.1.2
    pip install requests==2.31.0
    pip install tqdm==4.66.1
    pip install importlib-metadata==4.8.1
    pip install packaging==23.2
    pip install pydantic==1.10.13
    pip install sounddevice==0.4.6
    pip install python-multipart==0.0.6
    pip install cutlet>=0.3.0
    pip install unidic-lite>=1.0.8

They **should** all be playing nicely together... I believe, but if you want to check and let me know where you get to.

**Reply by erew123** [reply]
I should add.. Im happy to relax the requirements... its more a case of finding a way to get the resolver to say everything is in balance when you have torch with CUDA installed.

**Comment by erew123** [comment]
There's possible 1x extra complexity to add to this now... https://github.com/erew123/alltalk_tts/issues/82

Huggingface have updated transformers and there may be a bug or change that has broken things.

**Reply by erew123** [reply]
oops. didnt mean to close this!

**Comment by stratus-ss** [comment]
I see JoshXT beat me to it. This is not relevant anymore

**Reply by erew123** [reply]
Hiya! Yeah, I was just looking at this. I'm seeing if it works. I thought I would try that before sending you a message. Thanks and of course apologies. I kind of feel bad, but I guess I didn't know someone else was going to submit a PR with it done. Thanks for your time on it though  :)

**Reply by stratus-ss** [reply]
no issue, JoshXT is well known in the community. 

I'll give his container a spin. I looked it over and the only comment I have is that his docker-compose files are pulling from his org on dockerhub. Easily changeable. 

You asked about how changes work I can clarify

Using the docker-compose files **will not** pick up your changes as it stands now. It is referencing a pre-built container. It would be more advisable to have a small shell script that will build the Dockerfile locally and reference that in the docker-compose. This way the user  can always have the latest container.

One way I have seen this is to set a var in the shell script that is something like `BUILD_CONTAINER=true` or similar and have that set. This way it always rebuilds the container from your sources to be current. If set to false it could reference a specific container version hosted in your org on Dockerhub or Quay.io or wherever you choose to host your containers.

I work with containers every day, happy to help/advise

**Reply by erew123** [reply]
I think I've got the org changed correctly, everything is in dev https://github.com/erew123/alltalk_tts/tree/dev at the moment (changed 3x files).

I'm taking this a bit slowly and trying to learn on the way. When you say "set a var in the shell script" do you mean in the docker scripts? (sorry still learning) Also I note that https://github.com/erew123/alltalk_tts/blob/dev/.github/workflows/publish-docker.yml says Ubuntu-latest, so I'm assuming this means the docker container will be Linux only currently.

Maybe the best thing is that I at least get this initial stage working, get a base knowledge and then ask questions, as I am probably asking things that are relatively simple, but a pain in the ass to explain on text when I have little knowledge! So maybe I should get back to you with questions later....

**Reply by stratus-ss** [reply]
Actually my job is to explain these things to clients all day every day. If you have a place where we could DM, I am more than happy to field any and all questions you have. We were all new to something at one point. I'm happy to get on a voice call of some sort as well

**Reply by erew123** [reply]
Sure Id be happy to chat if you have 5-10 minutes. Im just thinking of a decent way of doing that or maybe swapping a phone number that doesn't compromise leaving any of our personal info public. If you have discord, I'm on there as kudos76 and I've just updated my avatar image to be the same as on here. Happy to swap numbers on there if you like or if you have any other suggestions?
